File: src-tauri/src/app/helpers/semver.rs
use crate::app::modules::{manifest::models::VersionSpec, repositories::models::VersionResult};
use semver::{Version, VersionReq};

pub fn is_semver_range(s: &str) -> bool {
    matches!(s.chars().next(), Some('^' | '~' | '>' | '<' | '*'))
}

pub fn resolve_version(range: &str, available: &[VersionResult]) -> Option<VersionResult> {
    resolve_with_semver(range, available)
        .or_else(|| resolve_with_normalized(range, available))
        .or_else(|| resolve_with_exact(range, available))
}

pub fn satisfies(spec: &VersionSpec, version: &str) -> bool {
    match spec {
        VersionSpec::Exact(v) => v == version,
        VersionSpec::Range(r) => resolve_with_semver(
            r,
            &[VersionResult {
                mod_id: String::new(),
                version: version.to_string(),
                minecraft_versions: vec![],
                url: String::new(),
                hash: String::new(),
            }],
        )
        .is_some(),
    }
}

/// Normalize a version string by stripping leading non-numeric characters
/// Example: "mc1.21.1-0.6.5-fabric" → "0.6.5-fabric"
fn normalize_version(s: &str) -> String {
    let mut chars = s.chars();
    let mut result = String::new();
    let mut found_digit = false;

    while let Some(c) = chars.next() {
        if !found_digit && !c.is_ascii_digit() {
            continue;
        }
        found_digit = true;
        result.push(c);
    }

    result
}

fn normalize_semver_string(s: &str) -> String {
    // keep prefix characters (~,^,>,<,=) if present
    let mut prefix = String::new();
    let mut core_start = 0;
    for (i, c) in s.char_indices() {
        if c.is_ascii_digit() {
            core_start = i;
            break;
        }
        prefix.push(c);
    }

    let core = &s[core_start..];
    let mut parts: Vec<&str> = core.split('.').collect();

    if parts.len() > 3 {
        parts.truncate(3); // drop extra build digits
    } else if parts.len() < 3 {
        while parts.len() < 3 {
            parts.push("0");
        }
    }

    format!("{}{}", prefix, parts.join("."))
}

/// Try resolving with raw semver
fn resolve_with_semver(range: &str, available: &[VersionResult]) -> Option<VersionResult> {
    let normalized = normalize_semver_string(range);
    let req = VersionReq::parse(&normalized).ok()?;
    let mut parsed: Vec<_> = available
        .iter()
        .filter_map(|v| {
            let fixed = normalize_semver_string(&v.version);
            Version::parse(&fixed).ok().map(|sv| (sv, v.clone()))
        })
        .collect();
    parsed.sort_by(|a, b| b.0.cmp(&a.0));
    parsed
        .into_iter()
        .find(|(sv, _)| req.matches(sv))
        .map(|(_, v)| v)
}

/// Try resolving with normalized semver
fn resolve_with_normalized(range: &str, available: &[VersionResult]) -> Option<VersionResult> {
    let r_norm = normalize_version(range);

    let available_norm: Vec<VersionResult> = available
        .iter()
        .map(|v| {
            let mut clone = v.clone();
            clone.version = normalize_version(&clone.version);
            clone
        })
        .collect();

    resolve_with_semver(&r_norm, &available_norm)
}

/// Fallback: try exact match without range operators
fn resolve_with_exact(range: &str, available: &[VersionResult]) -> Option<VersionResult> {
    let clean = range.trim_start_matches(|c: char| "^~><=*".contains(c));
    available.iter().find(|v| v.version == clean).cloned()
}

File: packages/core/src/helpers/semver.ts
// Semver resolution ported from src-tauri/src/app/helpers/semver.rs
// 3-tier fallback: raw semver → normalized semver → exact match

import semver from "semver";
import type { VersionSpec } from "../models/manifest.js";
import type { VersionResult } from "../models/repository.js";

export { isSemverRange } from "../models/manifest.js";

/**
 * Resolve a version spec against available versions.
 * Tries 3 strategies in order: semver, normalized semver, exact match.
 */
export function resolveVersion(range: string, available: VersionResult[]): VersionResult | undefined {
  return (
    resolveWithSemver(range, available) ??
    resolveWithNormalized(range, available) ??
    resolveWithExact(range, available)
  );
}

/** Check if a version satisfies a VersionSpec */
export function satisfies(spec: VersionSpec, version: string): boolean {
  if (spec.kind === "exact") {
    return spec.value === version;
  }
  // For range, try resolving the single version against the range
  const dummy: VersionResult = {
    modId: "",
    version,
    minecraftVersions: [],
    url: "",
    hash: "",
  };
  return resolveWithSemver(spec.value, [dummy]) !== undefined;
}

/** Compare two version strings. Returns <0, 0, or >0. */
export function compareVersions(a: string, b: string): number {
  const na = normalizeVersion(a);
  const nb = normalizeVersion(b);
  const va = semver.valid(semver.coerce(na));
  const vb = semver.valid(semver.coerce(nb));

  if (va && vb) return semver.compare(va, vb);
  if (va && !vb) return 1; // valid > invalid
  if (!va && vb) return -1;
  return a < b ? -1 : a > b ? 1 : 0; // fallback: lexicographic
}

/**
 * Strip leading non-digit characters from a version string.
 * Example: "mc1.21.1-0.6.5-fabric" → "1.21.1-0.6.5-fabric"
 *
 * Note: The Rust implementation strips everything until the first digit,
 * which for "mc1.21.1-0.6.5-fabric" gives "1.21.1-0.6.5-fabric" (starting at '1').
 */
function normalizeVersion(s: string): string {
  let i = 0;
  while (i < s.length && !/\d/.test(s[i])) {
    i++;
  }
  return s.slice(i);
}

/**
 * Normalize a semver string by keeping prefix chars (^, ~, >, <, =)
 * and ensuring exactly 3 numeric parts.
 */
function normalizeSemverString(s: string): string {
  // Keep prefix characters
  let prefixEnd = 0;
  for (let i = 0; i < s.length; i++) {
    if (/\d/.test(s[i])) {
      prefixEnd = i;
      break;
    }
    prefixEnd = i + 1;
  }

  const prefix = s.slice(0, prefixEnd);
  const core = s.slice(prefixEnd);
  const parts = core.split(".");

  if (parts.length > 3) {
    parts.length = 3; // truncate
  } else {
    while (parts.length < 3) {
      parts.push("0");
    }
  }

  return prefix + parts.join(".");
}

/** Try resolving with raw semver (tier 1) */
function resolveWithSemver(range: string, available: VersionResult[]): VersionResult | undefined {
  const normalizedRange = normalizeSemverString(range);
  const req = semver.validRange(normalizedRange);
  if (!req) return undefined;

  const parsed: Array<{ sv: string; original: VersionResult }> = [];
  for (const v of available) {
    const fixed = normalizeSemverString(v.version);
    const sv = semver.valid(semver.coerce(fixed));
    if (sv) {
      parsed.push({ sv, original: v });
    }
  }

  // Sort descending (highest first)
  parsed.sort((a, b) => semver.rcompare(a.sv, b.sv));

  // Find first that satisfies the range
  const match = parsed.find((p) => semver.satisfies(p.sv, req));
  return match?.original;
}

/** Try resolving with normalized versions (tier 2) */
function resolveWithNormalized(range: string, available: VersionResult[]): VersionResult | undefined {
  const normalizedRange = normalizeVersion(range);

  // Create normalized copies of available versions but keep original references
  const normalizedAvailable: VersionResult[] = available.map((v) => ({
    ...v,
    version: normalizeVersion(v.version),
  }));

  const result = resolveWithSemver(normalizedRange, normalizedAvailable);
  if (!result) return undefined;

  // Find the original (un-normalized) version that corresponds to this result
  // Match by URL since it's unique per version
  return available.find((v) => v.url === result.url) ?? result;
}

/** Fallback: try exact match without range operators (tier 3) */
function resolveWithExact(range: string, available: VersionResult[]): VersionResult | undefined {
  const clean = range.replace(/^[\^~><=*]+/, "");
  return available.find((v) => v.version === clean);
}
